postgis/lwgeom_functions_basic.c

```diff
@@ -2399,6 +2400,129 @@ Datum LWGEOM_azimuth(PG_FUNCTION_ARGS)
 	PG_RETURN_FLOAT8(result);
 }
 
+/**
+ * Compute the angle defined by 3 points or the angle between 2 vectors
+ * defined by 4 points
+ * given Point geometries.
+ * @return NULL on exception (same point).
+ * 		Return radians otherwise (always positive).
+ */
+PG_FUNCTION_INFO_V1(LWGEOM_angle);
+Datum LWGEOM_angle(PG_FUNCTION_ARGS)
+{
+	GSERIALIZED * seri_geoms[4];
+	LWGEOM *geom_unser;
+	LWPOINT *lwpoint;
+	POINT2D points[4]; 
+	double az1,az2 ;
+	double result;
+	int srids[4];
+	int i = 0 ; 
+	int j = 0;
+	int err_code = 0; 
+	int n_args = PG_NARGS(); 
+
+	/* no deserialize, checking for common error first*/
+	for(i=0; i<n_args; i++)
+	{
+		seri_geoms[i] = PG_GETARG_GSERIALIZED_P(i); 
+		if (gserialized_is_empty(seri_geoms[i]) )
+		{/* empty geom */
+			if (i==3)
+			{
+				n_args = 3 ;
+			}
+			else
+			{
+				err_code = 1 ;
+				break ;
+			}
+		} else
+		{
+			if(gserialized_get_type(seri_geoms[i]) != POINTTYPE)
+			{/* geom type */
+				err_code = 2 ;
+				break;
+			}
+			else
+			{
+				srids[i] = gserialized_get_srid(seri_geoms[i]) ; 
+				if(srids[0] != srids[i])
+				{/* error on srid*/
+					err_code = 3 ;
+					break;
+				}
+			}
+		}
+	}
+	if (err_code >0)
+		switch (err_code){
+			default: /*always executed*/
+			for (j=0;j<=i;j++)
+				PG_FREE_IF_COPY(seri_geoms[j], j);
+			
+			case 1:
+			lwpgerror("Empty geometry"); 
+			PG_RETURN_NULL() ;
+			break;
+			
+			case 2:
+			lwpgerror("Argument must be POINT geometries");
+			PG_RETURN_NULL();
+			break;
+			
+			case 3:
+			lwpgerror("Operation on mixed SRID geometries");
+			PG_RETURN_NULL();
+			break;
+		}
+	/* extract points */
+	for(i=0; i<n_args; i++)
+	{ 
+		geom_unser = lwgeom_from_gserialized(seri_geoms[i]) ; 
+		lwpoint = lwgeom_as_lwpoint(geom_unser); 
+		if (!lwpoint)
+		{ 
+			for (j=0;j<n_args;j++)
+				PG_FREE_IF_COPY(seri_geoms[j], j);
+			lwpgerror("Error unserializing geometry"); 
+			PG_RETURN_NULL() ;
+		} 
+		
+		if ( ! getPoint2d_p(lwpoint->point, 0, &points[i]) )
+		{ 
+			for (j=0;j<n_args;j++)
+				PG_FREE_IF_COPY(seri_geoms[j], j);
+			lwpgerror("Error extracting point");
+			PG_RETURN_NULL();
+		}
+		lwfree(geom_unser);
+		/* dont do lwpoint_free(lwpoint); , this memory is needed ! */ 
+	}
+	for (j=0;j<n_args;j++)
+		PG_FREE_IF_COPY(seri_geoms[j], j);
```

Comment:
It is not safe to release the serialized format after deserializing, because LWGEOM can still refer to the serialized POINTLIST.